crates/typst-cli/src/eval.rs

```diff
@@ -0,0 +1,148 @@
+            .map(|output| output.map(|document| document.introspector)),
+    };
+
+    match output {
+        // Retrieve and print evaluation results.
+        Ok(introspector) => {
+            let scope = evaluate_scope(&command.scope, &world, &introspector)?;
+            let statement = match &command.statement {
+                StringInput::Stdin => read_statement_from_stdin()?,
+                StringInput::String(statement) => statement.clone(),
+            };
+            let eval_result =
+                evaluate_statement(statement, command.mode, scope, &world, &introspector);
+            let errors = match &eval_result {
+                Err(errors) => errors.as_slice(),
+                Ok(value) => {
+                    let serialized =
+                        crate::serialize(value, command.format, command.pretty)?;
+                    println!("{serialized}");
+                    &[]
+                }
+            };
+            print_diagnostics(
+                &world,
+                errors,
+                &warnings,
+                command.process.diagnostic_format,
+            )
+            .map_err(|err| eco_format!("failed to print diagnostics ({err})"))?;
+        }
+
+        // Print diagnostics.
+        Err(errors) => {
+            set_failed();
+            print_diagnostics(
+                &world,
+                &errors,
+                &warnings,
+                command.process.diagnostic_format,
+            )
+            .map_err(|err| eco_format!("failed to print diagnostics ({err})"))?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Evaluates the scope with values interpreted as Typst code.
+fn evaluate_scope(
+    key_value_pairs: &[(String, String)],
+    world: &dyn World,
+    introspector: &Introspector,
+) -> HintedStrResult<Scope> {
+    let mut scope = Scope::new();
+
+    for (key, value) in key_value_pairs {
+        let value = evaluate_statement(
+            value.clone(),
+            SyntaxMode::Code,
+            Scope::default(),
+            world,
+            introspector,
+        )
+        .map_err(|errors| {
+            let mut message =
+                EcoString::from(format!("failure in scope key `{key}` evaluation"));
+            for (i, error) in errors.into_iter().enumerate() {
+                message.push_str(if i == 0 { ": " } else { ", " });
+                message.push_str(&error.message);
+            }
+            message
+        })?;
+
+        scope.bind(key.into(), Binding::detached(value));
+    }
+
+    Ok(scope)
+}
+
+/// Evaluates the statement with the given mode and scope.
+fn evaluate_statement(
+    statement: String,
+    mode: SyntaxMode,
+    scope: Scope,
+    world: &dyn World,
+    introspector: &Introspector,
+) -> SourceResult<Value> {
+    eval_string(
+        &typst::ROUTINES,
+        world.track(),
+        // TODO: propagate warnings
+        Sink::new().track_mut(),
```

Comment:
Retaining the warnings here and printing them should be _fairly_ straightforward I think and might be good to have